core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -1293,24 +1416,51 @@ private RecordCursorFactory generateSampleBy(QueryModel model, SqlExecutionConte
                         functionParser,
                         executionContext,
                         groupByFunctions,
+                        groupByFunctionPositions,
                         valueTypes
                 );
 
                 final ObjList<Function> recordFunctions = new ObjList<>(columnCount);
                 final GenericRecordMetadata groupByMetadata = new GenericRecordMetadata();
+
                 GroupByUtils.prepareGroupByRecordFunctions(
                         model,
                         metadata,
                         listColumnFilterA,
                         groupByFunctions,
+                        groupByFunctionPositions,
                         recordFunctions,
+                        recordFunctionPositions,
                         groupByMetadata,
                         keyTypes,
                         valueTypes.getColumnCount(),
                         false,
                         timestampIndex
                 );
 
+
+                boolean isFillNone = fillCount == 0 || fillCount == 1 && Chars.equalsLowerCaseAscii(sampleByFill.getQuick(0).token, "none");
+                boolean allGroupsFirstLast = isFillNone && allGroupsFirstLastWithSingleSymbolFilter(model, metadata);
+                if (allGroupsFirstLast) {
+                    SingleSymbolFilter symbolFilter = factory.convertToSampleByIndexDataFrameCursorFactory();
+                    if (symbolFilter != null) {
+                        return new SampleByFirstLastRecordCursorFactory(
+                                factory,
+                                timestampSampler,
+                                groupByMetadata,
+                                model.getColumns(),
+                                metadata,
+                                timezoneNameFunc,
+                                timezoneNameFuncPos,
+                                offsetFunc,
+                                offsetFuncPos,
+                                timestampIndex,
+                                symbolFilter,
+                                configuration.getSampleByIndexSearchPageSize()
+                        );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource created by the following constructor: *SampleByFirstLastRecordCursorFactory*. The resource is returned at line 1447. There are other execution paths that don&#39;t close the resource or return it. To prevent this resource leak, close the object returned by *SampleByFirstLastRecordCursorFactory()* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).